src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -206,136 +219,134 @@ public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionControlle
-                                                                                           RangeIterator<PrimaryKey> it = buildIterator(op, e.getValue(), isAnnHybridSearch);
-                                                                                           if (isAnnHybridSearch)
-                                                                                               it = reorderAndLimitBy(it, e.getKey(), annExpressionInHybridSearch);
-                                                                                           return it;
-                                                                                       })
-                                                                                       .collect(Collectors.toList());
-
-            Iterable<RangeIterator<PrimaryKey>> allIntersections = Iterables.concat(sstableIntersections, memtableIntersections);
-
-            queryContext.sstablesHit += queryView.referencedIndexes
-                                        .stream()
-                                        .map(SSTableIndex::getSSTable).collect(Collectors.toSet()).size();
-            queryContext.checkpoint();
-            RangeIterator<PrimaryKey> union = RangeUnionIterator.build(allIntersections);
-            return new CheckpointingIterator<>(union,
-                                               queryView.referencedIndexes,
-                                               annQueryViewInHybridSearch == null ? Collections.emptySet() : annQueryViewInHybridSearch.referencedIndexes,
-                                               queryContext);
+            return TermIterator.build(sstableIntersections, memtableResults, queryView.referencedIndexes, queryContext);
         }
         catch (Throwable t)
         {
             // all sstable indexes in view have been referenced, need to clean up when exception is thrown
             queryView.referencedIndexes.forEach(SSTableIndex::release);
-            // if ANN sstable indexes are referenced separately, release them
-            if (annQueryViewInHybridSearch != null)
-                annQueryViewInHybridSearch.referencedIndexes.forEach(SSTableIndex::release);
-
             throw t;
         }
     }
 
-    private RangeIterator<PrimaryKey> convertToPrimaryKeyIterator(PrimaryKeyMap.Factory pkFactory, RangeIterator<Long> sstableRowIdsIterator)
+    // This is a hybrid query. We apply all other predicates before ordering and limiting.
+    public RangeIterator getTopKRows(RangeIterator source, RowFilter.Expression expression)
     {
-        if (sstableRowIdsIterator.getCount() <= 0)
-            return RangeIterator.emptyKeys();
-
-        PrimaryKeyMap primaryKeyMap = pkFactory.newPerSSTablePrimaryKeyMap();
-        return SSTableRowIdKeyRangeIterator.create(primaryKeyMap, queryContext, sstableRowIdsIterator);
+        return new OrderingFilterRangeIterator(source,
+                                               ORDER_CHUNK_SIZE,
```

Comment:
I know what a big fan you are of the injection for globalBruteForceRows, but could we use a similar approach to make sure we're testing multiple loops of the chunking?

Reply:
:) I had a test written based on similar logic for `INTERSECTION_CLAUSE_LIMIT` in `SelectiveIntersectionTest`. I couldn't get the test to fail on any of my earlier commits, so I didn't commit it. I can add it back though.

Reply:
Your call, I'm okay either way.